### PR TITLE
Setup for newer operator routines which have large and small float types

### PR DESCRIPTION
**Pull request by cjolivier01**
BatchNorm (and I presume others) have accumulators, for instance, which may require a larger type to hold summations or other progressive operations

### CODE REVIEW

**Comment by cjolivier01** [comment]
Some Torch algorithms tend to have both types

File: mshadow/base.h [hunk 1]
```diff
@@ -678,6 +678,41 @@ struct minimum {
     LOG(FATAL) << "Unknown type enum " << type;       \
   }
 
+#define MSHADOW_REAL_TYPE_SWITCH_EX(type, DType, DLargeType, ...)  \
+  switch (type) {                                   \
+  case mshadow::kFloat32:                           \
+    {                                               \
+      typedef float DType;                          \
+      typedef double DLargeType;                    \
```
Comment (piiswrong) [inline mshadow/base.h]: This is not needed.

File: mshadow/base.h [hunk 2]
```diff
@@ -678,6 +678,41 @@ struct minimum {
...
+    {                                               \
+      typedef double DType;                         \
+      typedef double DLargeType;                    \
+      {__VA_ARGS__}                                 \
+    }                                               \
+    break;                                          \
+  case mshadow::kFloat16:                           \
+    {                                               \
+      typedef mshadow::half::half_t DType;          \
+      typedef float DLargeType;                     \
```
Comment (piiswrong) [inline mshadow/base.h]: This is sometimes usefull but may break backward compatibility.

File: mshadow/base.h (hunk 1 above)
Comment (piiswrong) [reply mshadow/base.h]: float is enough here

Comment (cjolivier01) [reply mshadow/base.h]: in Torch, they use double for summations across a tensor of floats in order to avoid overflows

File: mshadow/base.h (hunk 2 above)
Comment (cjolivier01) [reply mshadow/base.h]: Not sure what you mean -- this isn't use anywhere else so far, although I expect further non-mshadow ports to use it.  What backward compatibility might be broken?

**Pull request merged**